src/gradients.rs

```diff
@@ -52,7 +54,15 @@ impl<E: Unit, D: DeviceStorage> Gradients<E, D> {
 
     /// Drops all gradients except for the ids specified in the parameter.
     pub fn retain(&mut self, ids: &[UniqueId]) {
-        self.gradient_by_id.retain(|k, _| ids.contains(k));
+        self.non_temporary.extend(ids);
```

Comment:
non_temporary needs to be a set so this doesn't keep accumulating the same ids. A set may also be faster to query!